src/backend/wayland/mod.rs

```diff
@@ -1 +1,232 @@
+    }
+
+    // spin util we got all outputs from wayland
+    await_queue_events(
+        |state| state.outputs_fetched,
+        queue_mutex.clone(),
+        state_mutex.clone(),
+    )
+    .await?;
+
+    let screenshot_manager = {
+        let state = &mut *state_mutex.lock().unwrap();
+
+        state
+            .zwlr_screencopy_manager_v1
+            .as_ref()
+            .ok_or(WaylandError::NoScreenshotManager)?
+            .clone()
+    };
+
+    let num_of_outputs = {
+        let state = &mut *state_mutex.lock().unwrap();
+        state.outputs.len()
+    };
+
+    let mut screenshots: Vec<(OutputInfo, DynamicImage)> = vec![];
+    for i in 0..num_of_outputs {
+        let output = {
+            let state = &mut *state_mutex.lock().unwrap();
+            &state.outputs[i].clone()
+        };
+
+        let frame = screenshot_manager.capture_output(0, &output.output, &queue_handle, ());
+
+        // spin until screenshot ready
+        await_queue_events(
+            |state| state.current_frame.is_some(),
+            queue_mutex.clone(),
+            state_mutex.clone(),
+        )
+        .await?;
+
+        let (width, height, stride, format) = {
+            let state = &mut *state_mutex.lock().unwrap();
+
+            let current_frame = state
+                .current_frame
+                .as_ref()
+                .ok_or(WaylandError::BrokenState("current_frame"))?;
+
+            let format = current_frame.format.ok_or(WaylandError::MissingFormat)?;
+
+            let width = current_frame.width;
+            let height = current_frame.height;
+            let stride = current_frame.stride;
+
+            (width, height, stride, format)
+        };
+
+        let mut shared_memory = {
+            let state = &mut *state_mutex.lock().unwrap();
+
+            WaylandSharedMemory::new(
+                state.wl_shm.as_ref().ok_or(WaylandError::NoShmBind)?,
+                &queue_handle,
+                width,
+                height,
+                stride,
+                format,
+            )?
+        };
+
+        frame.copy(shared_memory.get_buffer());
+
+        // spin until screenshot copied into shared buffer
+        await_queue_events(
+            |state| state.screenshot_ready,
+            queue_mutex.clone(),
+            state_mutex.clone(),
+        )
+        .await?;
+
+        // read from shared memory
+        // data holds our screenshot
+        let mut data = vec![];
+        shared_memory.get_memfile().read_to_end(&mut data)?;
+
+        let output_info = OutputInfo::try_from({
+            let state = &mut *state_mutex.lock().unwrap();
+            &state.outputs[i].clone()
+        })?;
+        let img = image_from_wayland(data, width, height, format)?;
+
+        screenshots.push((output_info, img));
+
+        shared_memory.destroy();
+
+        let state = &mut *state_mutex.lock().unwrap();
+        state.next_screen(); // reset current screenshot metadata
```

Comment:
Ok, so in general I'd change the code format a bit in the loop to
get a better overview of what is happening in the loop:
```suggestion
        let img = {
            let output = {
                let state = &mut *state_mutex.lock().unwrap();
                &state.outputs[i].clone()
            };

            let frame = screenshot_manager.capture_output(0, &output.output, &queue_handle, ());

            // spin until screenshot ready
            await_queue_events(
                |state| state.current_frame.is_some(),
                queue_mutex.clone(),
                state_mutex.clone(),
            )
            .await?;

            let (width, height, stride, format) = {
                let state = &mut *state_mutex.lock().unwrap();

                let current_frame = state
                    .current_frame
                    .as_ref()
                    .ok_or(WaylandError::BrokenState("current_frame"))?;

                let format = current_frame.format.ok_or(WaylandError::MissingFormat)?;

                let width = current_frame.width;
                let height = current_frame.height;
                let stride = current_frame.stride;

                (width, height, stride, format)
            };

            let img_bytes = {
                let img_bytes = vec![];
                let mut shared_memory = {
                    let state = &mut *state_mutex.lock().unwrap();

                    WaylandSharedMemory::new(
                        state.wl_shm.as_ref().ok_or(WaylandError::NoShmBind)?,
                        &queue_handle,
                        width,
                        height,
                        stride,
                        format,
                    )?
                };

                frame.copy(shared_memory.get_buffer());

                // spin until screenshot copied into shared buffer
                await_queue_events(
                    |state| state.screenshot_ready,
                    queue_mutex.clone(),
                    state_mutex.clone(),
                )
                .await?;

                shared_memory.get_memfile().read_to_end(&mut img_bytes)?;
                shared_memory.destroy();

                img_bytes
            };

            image_from_wayland(img_bytes, width, height, format)?
        };

        let output_info = OutputInfo::try_from({
            let state = &mut *state_mutex.lock().unwrap();
            &state.outputs[i].clone()
        })?;

        screenshots.push((output_info, img));

        let state = &mut *state_mutex.lock().unwrap();
        state.next_screen(); // reset current screenshot metadata
```